lametro/views.py

```diff
@@ -379,54 +378,52 @@ def day_grouper(x):
-            org_all_events = []
+            org_all_events = itertools.groupby(all_events, key=day_grouper)
 
-            for event_date, events in itertools.groupby(all_events, key=day_grouper):
-                events = sorted(events, key=attrgetter("start_time"))
-                org_all_events.append([date(*event_date), events])
+            context["all_events"] = [(d, list(events)) for d, events in org_all_events]
 
-            context["all_events"] = org_all_events
-        # If no...
         else:
             # Upcoming events
             future_events = media_events.filter(start_time__gt=timezone.now()).order_by(
                 "start_time"
             )
-            org_future_events = []
-
-            for event_date, events in itertools.groupby(future_events, key=day_grouper):
-                events = sorted(events, key=attrgetter("start_time"))
-                org_future_events.append([date(*event_date), events])
-
-            context["future_events"] = org_future_events
+            org_future_events = itertools.groupby(future_events, key=day_grouper)
 
             # Past events
-            past_events = media_events.filter(start_time__lt=timezone.now()).order_by(
-                "-start_time"
+            past_events = (
+                media_events.filter(start_time__lt=timezone.now())
+                .order_by("-start_time")
+                .prefetch_related(
+                    Prefetch("documents", minutes_queryset, to_attr="minutes")
+                )
+                .prefetch_related("minutes__links")
             )
-
-            past_events = past_events.prefetch_related(
-                Prefetch("documents", minutes_queryset, to_attr="minutes")
-            ).prefetch_related("minutes__links")
-
-            org_past_events = []
-
-            for event_date, events in itertools.groupby(past_events, key=day_grouper):
-                events = sorted(events, key=attrgetter("start_time"))
-                org_past_events.append([date(*event_date), events])
-
-            context["past_events"] = org_past_events
+            org_past_events = itertools.groupby(past_events, key=day_grouper)
+
+            # Only show three days of events by default
+            past_events = itertools.islice(org_past_events, 3)
+            future_events = itertools.islice(org_future_events, 3)
+
+            # Did user request all past or future events?
+            show_all_past_events = self.request.GET.get("show") == "past"
+            show_all_future_events = self.request.GET.get("show") == "future"
+            if show_all_past_events:
+                past_events = org_past_events
+            elif show_all_future_events:
+                future_events = org_future_events
+
+            context["future_events"] = [
+                (d, list(events)) for d, events in future_events
+            ]
+            context["past_events"] = [(d, list(events)) for d, events in past_events]
```

Comment:
Rather than loading these into memory, is it possible to pass the generator to the template and iterate it there instead?

Reply:
That would be ideal but the template throws an error when passed the generators.

I think this is one of the reasons we should eventually move to a more standard pagination approach for the all past/future events view.